### ISSUE TITLE
Data::Dumper passes unprintable chars inside qr/.../ through to output

### ISSUE THREAD

**Issue by jimav** [Perl/perl5#21608]
(Note: The "Issues" link at https://metacpan.org/pod/Data::Dumper points here, not to a module-specific issues tracker)

Module: Data::Dumper

**Description**
If an octet which is not valid ASCII is in a compiled regex, Data::Dumper just outputs the octet as-is not as an escape.   I'm guessing this can't be fixed without some complicated logic to ensure that the result will `eval` back to an equivalent regex.   On the other hand, maybe blindly converting non-ASCII-printable octets to escapes would be exactly right.

Note that `qr/\x{01}/` and `$s = "\x{01}"; qr/$s/` produce different, but presumably equivalant regexes.

**Steps to Reproduce**
```
#!/usr/bin/env perl
use strict; use warnings;
use open ':std', 'encoding(UTF-8)';
use Data::Dumper;
print "Data::Dumper::VERSION = $Data::Dumper::VERSION\n";

my $octets = "\x{01}\x{02}\x{03}\x{e2}\x{98}\x{ba}";
my $re = qr/^\Q${octets}\E$/;

"$octets" =~ $re && "X$octets" !~ $re && print "Regex worked as expected\n";

my $qq0_output = Data::Dumper->new([$re])->Terse(1)->Useqq(0)->Dump;
my $qq1_output = Data::Dumper->new([$re])->Terse(1)->Useqq(1)->Dump;

#print "Useqq(0):$qq0_output\n"; # trashes my terminal
{ if (open my $pipe, "| od -t x1 -t c") {
    print "Useqq(0):\n"; stdout->flush;
    print $pipe $qq0_output; close $pipe;
  }
}
#print "Useqq(1):$qq1_output\n"; # trashes my terminal
{ if (open my $pipe, "| od -t x1 -t c") {
    print "Useqq(1):\n"; stdout->flush;
    print $pipe $qq1_output; close $pipe;
  }
}
foreach (split //,$qq1_output) {
  printf "Un-printable character in Useqq(1) result: 0x%02x\n", ord($_)
    unless /[[:print:]\r\n]/;
}
print "---done---\n";
```
Output:
```
Data::Dumper::VERSION = 2.183
Regex worked as expected
Useqq(0):
0000000  71  72  2f  5e  5c  01  5c  02  5c  03  5c  c3  a2  5c  c2  98
          q   r   /   ^   \ 001   \ 002   \ 003   \ 303 242   \ 302 230
0000020  5c  c2  ba  24  2f  0a
          \ 302 272   $   /  \n
0000026
Useqq(1):
0000000  71  72  2f  5e  5c  01  5c  02  5c  03  5c  c3  a2  5c  c2  98
          q   r   /   ^   \ 001   \ 002   \ 003   \ 303 242   \ 302 230
0000020  5c  c2  ba  24  2f  0a
          \ 302 272   $   /  \n
0000026
Un-printable character in Useqq(1) result: 0x01
Un-printable character in Useqq(1) result: 0x02
Un-printable character in Useqq(1) result: 0x03
Un-printable character in Useqq(1) result: 0xe2
Un-printable character in Useqq(1) result: 0x98
Un-printable character in Useqq(1) result: 0xba
---done---
```

**Expected behavior**
Ok,  "expected" is too strong.  It would be great if D::D output was restricted to printable ASCII so it always could be printed to an encoding filehandle.

```
Data::Dumper  2.183
This is perl 5, version 36, subversion 0 (v5.36.0) built for x86_64-linux-gnu-thread-multi
(with 53 registered patches, see perl -V for more detail)
```

**Comment by demerphq** [comment]
> Note that `qr/\x{01}/` and `$s = "\x{01}"; qr/$s/` produce different, but presumably equivalant regexes.

For that codepoint yes, but generally that is not the case. `qr/a\x{7c}b/` and `$s="a\x{7c}b"; qr/$s/` produce different and not equivalent regexes. 

```
perl -le'my $pat = qr/a\x{7c}b/; my $s="a\x{7c}b"; my $pat2= qr/$s/; print "b"=~/$_/ ? "yes: $_" : "no: $_" for $pat, $pat2;'
no: (?^:a\x{7c}b)
yes: (?^:a|b)
```

I think that DD could in theory escape the non-printable characters in a pattern, but it cannot use quotemeta() to do it as that would escape things it should not.  FWIW, you can use re::regexp_pattern() to extract the pattern from a qr// and from there play with the logic that might be required. See `perldoc re`  for details.

**Comment by jkeenan** [comment]
> (Note: The "Issues" link at https://metacpan.org/pod/Data::Dumper points here, not to a module-specific issues tracker)

That's correct.  Data-Dumper is "blead-first," *i.e.,* maintained by Perl 5 Porters with revisions appearing first in production releases of Perl and then (hopefully) released to CPAN for applicability to earlier versions of `perl`.  So this is the correct place to file this report.  Thanks.

> 
> Module: Data::Dumper
> 
> **Description** If an octet which is not valid ASCII is in a compiled regex, Data::Dumper just outputs the octet as-is not as an escape. I'm guessing this can't be fixed without some complicated logic to ensure that the result will `eval` back to an equivalent regex. On the other hand, maybe blindly converting non-ASCII-printable octets to escapes would be exactly right.
> 
> Note that `qr/\x{01}/` and `$s = "\x{01}"; qr/$s/` produce different, but presumably equivalant regexes.
> 
> **Steps to Reproduce**
> 
> ```
> #!/usr/bin/env perl
> use strict; use warnings;
> use open ':std', 'encoding(UTF-8)';
> use Data::Dumper;
> print "Data::Dumper::VERSION = $Data::Dumper::VERSION\n";
> 
> my $octets = "\x{01}\x{02}\x{03}\x{e2}\x{98}\x{ba}";
> my $re = qr/^\Q${octets}\E$/;
> 
> "$octets" =~ $re && "X$octets" !~ $re && print "Regex worked as expected\n";
> 
> my $qq0_output = Data::Dumper->new([$re])->Terse(1)->Useqq(0)->Dump;
> my $qq1_output = Data::Dumper->new([$re])->Terse(1)->Useqq(1)->Dump;
> 
> #print "Useqq(0):$qq0_output\n"; # trashes my terminal
> { if (open my $pipe, "| od -t x1 -t c") {
>     print "Useqq(0):\n"; stdout->flush;
>     print $pipe $qq0_output; close $pipe;
>   }
> }
> #print "Useqq(1):$qq1_output\n"; # trashes my terminal
> { if (open my $pipe, "| od -t x1 -t c") {
>     print "Useqq(1):\n"; stdout->flush;
>     print $pipe $qq1_output; close $pipe;
>   }
> }
> foreach (split //,$qq1_output) {
>   printf "Un-printable character in Useqq(1) result: 0x%02x\n", ord($_)
>     unless /[[:print:]\r\n]/;
> }
> print "---done---\n";
> ```
> 
> Output:
> 
> ```
> Data::Dumper::VERSION = 2.183
> Regex worked as expected
> Useqq(0):
> 0000000  71  72  2f  5e  5c  01  5c  02  5c  03  5c  c3  a2  5c  c2  98
>           q   r   /   ^   \ 001   \ 002   \ 003   \ 303 242   \ 302 230
> 0000020  5c  c2  ba  24  2f  0a
>           \ 302 272   $   /  \n
> 0000026
> Useqq(1):
> 0000000  71  72  2f  5e  5c  01  5c  02  5c  03  5c  c3  a2  5c  c2  98
>           q   r   /   ^   \ 001   \ 002   \ 003   \ 303 242   \ 302 230
> 0000020  5c  c2  ba  24  2f  0a
>           \ 302 272   $   /  \n
> 0000026
> Un-printable character in Useqq(1) result: 0x01
> Un-printable character in Useqq(1) result: 0x02
> Un-printable character in Useqq(1) result: 0x03
> Un-printable character in Useqq(1) result: 0xe2
> Un-printable character in Useqq(1) result: 0x98
> Un-printable character in Useqq(1) result: 0xba
> ---done---
> ```
> 
> **Expected behavior** Ok, "expected" is too strong. It would be great if D::D output was restricted to printable ASCII so it always could be printed to an encoding filehandle.
> 
> ```
> Data::Dumper  2.183
> This is perl 5, version 36, subversion 0 (v5.36.0) built for x86_64-linux-gnu-thread-multi
> (with 53 registered patches, see perl -V for more detail)
> ```
